# Atividade Prática — QA com APIs (SWAPI)

## Versão do aluno — com scripts base

**Tempo estimado:** 60 minutos  
**Ferramenta:** Google Colab  
**Formato:** em grupos de **2 a 4 componentes** ou, se preferirem, **individualmente**.  
**Objetivo:** investigar respostas de uma API, ajustar scripts simples e responder às perguntas como um QA.

### Instruções
Nesta atividade, vocês **não precisam criar tudo do zero**.  
Os scripts já estão prontos como base. A tarefa do grupo será:

1. **Executar** os blocos de código
2. **Alterar apenas o que for pedido**
3. **Observar os resultados**
4. **Responder às perguntas em Markdown**
5. **Discutir em grupo** quais validações um QA faria

### Entrega
Cada grupo pode entregar **um único notebook** com:
- nomes dos integrantes
- respostas das perguntas
- scripts ajustados
- asserts completados

Se alguém optar por fazer sozinho, pode entregar individualmente sem problema.


In [1]:
import requests
import json

# -------------------------
# MODO DE SEGURANÇA (fallback)
# -------------------------
# Se a API não responder, o notebook usa dados locais
# para a atividade continuar funcionando.

DADOS_FALLBACK = {
    "https://swapi.dev/api/people/10/": {
        "name": "Obi-Wan Kenobi",
        "height": "182",
        "mass": "77",
        "eye_color": "blue-gray",
        "gender": "male",
        "url": "https://swapi.dev/api/people/10/"
    },
    "https://swapi.dev/api/people/13/": {
        "name": "Chewbacca",
        "height": "228",
        "mass": "112",
        "eye_color": "blue",
        "gender": "male",
        "url": "https://swapi.dev/api/people/13/"
    },
    "https://swapi.dev/api/planets/2/": {
        "name": "Alderaan",
        "climate": "temperate",
        "terrain": "grasslands, mountains",
        "population": "2000000000",
        "url": "https://swapi.dev/api/planets/2/"
    },
    "https://swapi.dev/api/starships/10/": {
        "name": "Millennium Falcon",
        "model": "YT-1300 light freighter",
        "manufacturer": "Corellian Engineering Corporation",
        "max_atmosphering_speed": "1050",
        "url": "https://swapi.dev/api/starships/10/"
    }
}

class RespostaLocal:
    def __init__(self, status_code, data):
        self.status_code = status_code
        self._data = data
        self.text = json.dumps(data, ensure_ascii=False)

    def json(self):
        return self._data

def buscar_endpoint(url):
    try:
        resposta = requests.get(url, timeout=10)
        if resposta.status_code == 200:
            return resposta, "online"
    except Exception:
        pass

    if url in DADOS_FALLBACK:
        return RespostaLocal(200, DADOS_FALLBACK[url]), "fallback local"

    return RespostaLocal(404, {"detail": "Not found"}), "fallback local"

print("Ambiente preparado com sucesso.")

Ambiente preparado com sucesso.


## Identificação do grupo

- Integrante 1: Anna Lívia Souza e Silva - RA: 18926

## Parte 1 — Explorando a API

Nesta primeira parte, vocês vão consultar 4 endpoints.  
O trabalho do grupo será **alterar apenas a URL** em cada bloco, executar e observar o retorno.


### 1.1 Personagem 10

**Tarefa:** execute o script e observe:
- status code
- nome
- altura
- cor dos olhos

In [6]:
url = "https://swapi.dev/api/people/32/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Qui-Gon Jinn",
    "height": "193",
    "mass": "89",
    "hair_color": "brown",
    "skin_color": "fair",
    "eye_color": "blue",
    "birth_year": "92BBY",
    "gender": "male",
    "homeworld": "https://swapi.dev/api/planets/28/",
    "films": [
        "https://swapi.dev/api/films/4/"
    ],
    "species": [],
    "vehicles": [
        "https://swapi.dev/api/vehicles/38/"
    ],
    "starships": [],
    "created": "2014-12-19T16:54:53.618000Z",
    "edited": "2014-12-20T21:17:50.375000Z",
    "url": "https://swapi.dev/api/people/32/"
}


**Respostas do aluno — Personagem 13**

- Status code: 200
- Nome do personagem: Quin-Gon Jinn
- Altura: 193
- Cor dos olhos: blue
- O JSON parece completo? Sim

### 1.2 Personagem 13

**Tarefa:** neste bloco, troque apenas o endpoint se necessário, execute e observe o retorno.

In [7]:
url = "https://swapi.dev/api/people/24/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Bossk",
    "height": "190",
    "mass": "113",
    "hair_color": "none",
    "skin_color": "green",
    "eye_color": "red",
    "birth_year": "53BBY",
    "gender": "male",
    "homeworld": "https://swapi.dev/api/planets/29/",
    "films": [
        "https://swapi.dev/api/films/2/"
    ],
    "species": [
        "https://swapi.dev/api/species/7/"
    ],
    "vehicles": [],
    "starships": [],
    "created": "2014-12-15T12:53:49.297000Z",
    "edited": "2014-12-20T21:17:50.355000Z",
    "url": "https://swapi.dev/api/people/24/"
}


**Respostas do aluno — Personagem 13**

- Status code: 200
- Nome do personagem: Bossk
- Altura: 190
- Cor dos olhos: red
- O JSON parece completo? Sim

### 1.3 Planeta 2

**Tarefa:** execute o bloco e observe:
- nome do planeta
- clima
- terreno
- população

In [8]:
url = "https://swapi.dev/api/planets/2/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Alderaan",
    "rotation_period": "24",
    "orbital_period": "364",
    "diameter": "12500",
    "climate": "temperate",
    "gravity": "1 standard",
    "terrain": "grasslands, mountains",
    "surface_water": "40",
    "population": "2000000000",
    "residents": [
        "https://swapi.dev/api/people/5/",
        "https://swapi.dev/api/people/68/",
        "https://swapi.dev/api/people/81/"
    ],
    "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/6/"
    ],
    "created": "2014-12-10T11:35:48.479000Z",
    "edited": "2014-12-20T20:58:18.420000Z",
    "url": "https://swapi.dev/api/planets/2/"
}


**Respostas do aluno — Planeta 2**

- Status code: 200
- Nome do planeta: Alderaan
- Clima: temperate
- Terreno: grasslands, mountains
- População: 2000000000
- O JSON parece completo? Sim

### 1.4 Nave 10

**Tarefa:** execute o bloco e observe:
- nome da nave
- modelo
- fabricante
- velocidade máxima em atmosfera

In [9]:
url = "https://swapi.dev/api/starships/10/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Millennium Falcon",
    "model": "YT-1300 light freighter",
    "manufacturer": "Corellian Engineering Corporation",
    "cost_in_credits": "100000",
    "length": "34.37",
    "max_atmosphering_speed": "1050",
    "crew": "4",
    "passengers": "6",
    "cargo_capacity": "100000",
    "consumables": "2 months",
    "hyperdrive_rating": "0.5",
    "MGLT": "75",
    "starship_class": "Light freighter",
    "pilots": [
        "https://swapi.dev/api/people/13/",
        "https://swapi.dev/api/people/14/",
        "https://swapi.dev/api/people/25/",
        "https://swapi.dev/api/people/31/"
    ],
    "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/"
    ],
    "created": "2014-12-10T16:59:45.094000Z",
    "edited": "2014-12-20T21:23:49.880000Z",
    "url": "https://swapi.dev/api/starships/10/"
}


**Respostas do aluno — Nave 10**

- Status code: 200
- Nome da nave: Millennium Falcon
- Modelo: YT-1300 light freighter
- Fabricante: Corellian Engineering Corporation
- Velocidade máxima em atmosfera: 1050
- O JSON parece completo? Sim

## Parte 2 — Extraindo campos específicos

Agora você vai executar os scripts e **completar as respostas**.

### 2.1 Extraindo campos do personagem

**Tarefa:** execute o bloco e depois responda.

In [10]:
url = "https://swapi.dev/api/people/10/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Nome:", dados["name"])
print("Altura:", dados["height"])
print("Cor dos olhos:", dados["eye_color"])
print("Gênero:", dados["gender"])

Nome: Obi-Wan Kenobi
Altura: 182
Cor dos olhos: blue-gray
Gênero: male


**Perguntas**
1. Qual campo foi usado para mostrar o nome? print("Nome:", dados["name"])
2. Qual campo foi usado para mostrar a altura? print("Altura:", dados["height"])
3. Se você fosse QA, quais desses campos validaria primeiro? Nome

### 2.2 Extraindo campos do planeta

**Tarefa:** execute o bloco e complete as respostas.

In [12]:
url = "https://swapi.dev/api/planets/2/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Nome do planeta:", dados["name"])
print("Clima:", dados["climate"])
print("Terreno:", dados["terrain"])
print("População:", dados["population"])

Nome do planeta: Alderaan
Clima: temperate
Terreno: grasslands, mountains
População: 2000000000


**Perguntas**
1. O campo população parece numérico? Sim
2. O clima foi retornado corretamente? Sim
3. Se você fosse QA, o que validaria nesse planeta? Se os campos estão preenchidos corretamente.

## Parte 3 — Pensando como QA

Agora você vai trabalhar com validações simples.

### 3.1 Validações básicas

**Tarefa:** execute o bloco abaixo e depois escreva o que cada validação verifica.

In [13]:
url = "https://swapi.dev/api/people/10/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Validação 1 - status 200:", response.status_code == 200)
print("Validação 2 - campo 'name' existe:", "name" in dados)
print("Validação 3 - nome não está vazio:", dados["name"] != "")
print("Validação 4 - campo 'height' existe:", "height" in dados)

Validação 1 - status 200: True
Validação 2 - campo 'name' existe: True
Validação 3 - nome não está vazio: True
Validação 4 - campo 'height' existe: True


**Explique com suas palavras**
1. O que a validação 1 verifica? Verifica os status do código, ou seja, valida a requisição foi bem-sucedida.
2. O que a validação 2 verifica? Verifica se o campo name existe
3. O que a validação 3 verifica? Verifica se o campo nome não está vazio
4. O que a validação 4 verifica? Verifica se o campo height existe

### 3.2 Complete as validações

**Tarefa:** execute o bloco abaixo. Ele já está funcionando. Depois, alterem as expressões se o professor pedir outra nave ou outro endpoint. Em seguida, expliquem por que cada validação faz sentido para QA.


In [14]:
url = "https://swapi.dev/api/starships/10/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Status 200:", response.status_code == 200)

# Os exemplos abaixo já estão corretos e executáveis.
# Se o professor pedir, vocês podem alterar para validar outros campos.
print("Campo 'name' existe:", "name" in dados)
print("Campo 'model' existe:", "model" in dados)
print("Nome não está vazio:", dados["name"] != "")


Status 200: True
Campo 'name' existe: True
Campo 'model' existe: True
Nome não está vazio: True


Status 200: True
O código 200 significa requisição bem-sucedida, garante que o endpoint respondeu corretamente.

Campo 'name' existe: True
Verifica se a estrutura do JSON está correta, confirma que a API não mudou o schema.

Campo 'model' existe: True
Confirma que informações importantes da nave existem.

Nome não está vazio: True
O campo existe, mas também precisa ter conteúdo válido.

## Parte 4 — Testes automatizados com assert

Agora vocês vão trabalhar sobre scripts já prontos. Os asserts abaixo já executam sem erro. A tarefa do grupo é **entender o que cada linha valida**, ajustar se o professor pedir outro recurso e registrar as respostas.


### 4.1 Assert no personagem

**Tarefa:** execute o bloco. Depois, expliquem o que cada assert valida e, se quiserem, alterem o endpoint para outro personagem.


In [24]:
url = "https://swapi.dev/api/people/12/"
response, origem = buscar_endpoint(url)
dados = response.json()

assert response.status_code == 200, "Erro: status inesperado"
assert "name" in dados, "Erro: campo name não encontrado"
assert dados["name"] != "", "Erro: nome vazio"

print("Todos os testes do personagem passaram!")


Todos os testes do personagem passaram!


assert response.status_code == 200,
Confirma que a requisição foi bem-sucedida.

assert "name" in dados,
Confirma que o JSON retornado possui o campo name.

assert dados["name"] != "",
Confirma que o campo name possui conteúdo.

### 4.2 Assert na nave

**Tarefa:** execute o bloco. Depois, descrevam qual campo foi validado em cada assert e proponham mais 1 validação extra em Markdown.


In [25]:
url = "https://swapi.dev/api/starships/15/"
response, origem = buscar_endpoint(url)
dados = response.json()

assert response.status_code == 200, "Erro: status inesperado"
assert "manufacturer" in dados, "Erro: fabricante não encontrado"
assert dados["model"] != "", "Erro: modelo vazio"

#Assert extra

assert dados["name"] == "Executor", "Erro: nome da nave inesperado"

print("Todos os testes da nave passaram!")

Todos os testes da nave passaram!


assert response.status_code == 200, Confirma que a API respondeu com status 200 (sucesso).

assert "manufacturer" in dados, Confirma que o campo fabricante da nave existe no JSON retornado.

assert dados["model"] != "", Confirma que o modelo da nave não está vazio.

assert dados["name"] == "Executor", Confirma se o campo name contém o nome esperado da nave.

## Parte 5 — Cenário negativo

Agora você vai observar como a API se comporta quando pedimos um recurso inexistente.

In [26]:
url = "https://swapi.dev/api/people/9999/"
response, origem = buscar_endpoint(url)

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print("Texto retornado:")
print(response.text)

Origem dos dados: fallback local
Status Code: 404
Texto retornado:
{"detail": "Not found"}


**Perguntas**
1. Qual status code foi retornado? 404
2. Esse comportamento está correto? Sim, pois o código 404 é retornado quando a requisição não foi encontrada.
3. Por que um QA deve testar cenários negativos? Garantir tratamento correto de erros e estabilidade da API

## Parte 6 — Mini desafio final

Agora você vai alterar apenas o endpoint da função abaixo e gerar um pequeno resumo.

In [28]:
def resumo_recurso(url):
    response, origem = buscar_endpoint(url)
    print("Origem dos dados:", origem)
    print("Status Code:", response.status_code)

    if response.status_code != 200:
        print("Recurso não encontrado.")
        return

    dados = response.json()

    for chave, valor in dados.items():
        print(f"{chave}: {valor}")

# ALTERE O ENDPOINT ABAIXO PARA TESTAR OUTRO RECURSO
resumo_recurso("https://swapi.dev/api/planets/7/")

Origem dos dados: online
Status Code: 200
name: Endor
rotation_period: 18
orbital_period: 402
diameter: 4900
climate: temperate
gravity: 0.85 standard
terrain: forests, mountains, lakes
surface_water: 8
population: 30000000
residents: ['https://swapi.dev/api/people/30/']
films: ['https://swapi.dev/api/films/3/']
created: 2014-12-10T11:50:29.349000Z
edited: 2014-12-20T20:58:18.429000Z
url: https://swapi.dev/api/planets/7/


**Respostas finais**
1. Qual endpoint você testou no desafio? 7 (https://swapi.dev/api/planets/7/)
2. Qual foi o status code? 200
3. Qual campo chamou mais a sua atenção? films
4. Se você fosse QA, quais 3 validações faria nesse recurso? Name, campos vazios e se os campos estão sendo preenchidos corretamente.

# Entrega da atividade

Entreguem:

1. O notebook preenchido  
2. As respostas em Markdown  
3. Os scripts ajustados  
4. Os asserts analisados e/ou adaptados  
5. Uma conclusão curta do grupo: **quais foram as 3 validações mais importantes?**

## Fechamento
Ao final desta atividade, vocês praticaram:

- leitura de resposta JSON
- análise de status code
- validação de campos
- pensamento de QA
- testes simples com assert
- análise de cenário negativo
